records/2021-09-27.batch-celery-save.md

```diff
@@ -0,0 +1,116 @@
+# Database performance issues
+
+Date: 2021-09-27
+
+## Status
+
+**PROPOSED.**
+
+## Context
+
+Following the ADR around scalability this previous July, one of the recommendation
+is to perform a batch save of the notifications when these are incoming into our
```

Comment:
How do notifications come into the API? Is it a series of send messages, one per notification, or is it a single message which retrieves a list?

Reply:
We support 2 endpoints: bulk and atomic. The former is recent so not something that many clients use yet. The bulk endpoint would contain a series of message to send.

Reply:
Right, so bulk shouldn't be too bad as the messages are already grouped and you just push the single bulk message into the queue, yea?

For atomic, I'm guessing that means a single message to send a single notification, so if you want to send 10k messages you send 10k atomic API messages?

Maybe worth putting that distinction in here and spelling how what happens with bulk vs atomic messages (makes it easier for someone reading this without context in the future)

Reply:
> so bulk shouldn't be too bad as the messages are already grouped and you just push the single bulk message into the queue, yea?

Ideally, yes. Right now, every item in bulk is being sent individually into a queue and processed by one task. So a bulk with 1,000 rows will produce 1,000 task runs and 1,000 inserts in the database (at least). 

> For atomic, I'm guessing that means a single message to send a single notification, so if you want to send 10k messages you send 10k atomic API messages?

Yes.

> Maybe worth putting that distinction in here and spelling how what happens with bulk vs atomic messages (makes it easier for someone reading this without context in the future)

We can add more, these changes though will apply for both bulk and the atomic send endpoint.